File: qmd-python/src/qmd/cli/agent.py
"""Agent mode - interactive search with intelligent routing"""

import sys

from enum import Enum

from qmd.formatter.formatter import Formatter
from qmd.store.store import SearchOptions


class QueryIntent(str, Enum):
    """Query intent types"""
    KEYWORD = "keyword"  # BM25
    SEMANTIC = "semantic"  # Vector
    HYBRID = "hybrid"  # Hybrid


def classify_intent(query: str) -> QueryIntent:
    """Classify query intent using rule-based approach"""
    query_lower = query.lower()

    # Check for natural language patterns
    nl_patterns = ["explain", "describe", "what is", "how does", "why", "meaning"]
    for p in nl_patterns:
        if p in query_lower:
            return QueryIntent.SEMANTIC

    # Check for technical terms (likely keyword search)
    tech_patterns = ["error", "exception", "api", "function", "class", "method"]
    for p in tech_patterns:
        if p in query_lower:
            return QueryIntent.KEYWORD

    # Default to hybrid
    return QueryIntent.HYBRID


def run_agent(ctx):
    """Run interactive agent"""
    from qmd.cli.commands import get_store

    store = get_store(ctx.obj.get("db"))

    print("QMD Agent - Interactive Search")
    print("Type 'quit' or 'exit' to exit")
    print("Type '/bm25', '/vector', or '/hybrid' to force search method")
    print()

    force_method = ""

    while True:
        try:
            query = input("> ").strip()
        except (EOFError, KeyboardInterrupt):
            break

        if not query:
            continue

        if query.lower() in ("quit", "exit"):
            break

        # Check for forced method
        if query.startswith("/bm25"):
            force_method = "bm25"
            query = query.replace("/bm25", "").strip()
            print("Forced BM25 search")
        elif query.startswith("/vector"):
            force_method = "vector"
            query = query.replace("/vector", "").strip()
            print("Forced Vector search")
        elif query.startswith("/hybrid"):
            force_method = "hybrid"
            query = query.replace("/hybrid", "").strip()
            print("Forced Hybrid search")

        if not query:
            continue

        # Classify intent
        intent = classify_intent(query)

        if force_method:
            if force_method == "bm25":
                intent = QueryIntent.KEYWORD
            elif force_method == "vector":
                intent = QueryIntent.SEMANTIC
            elif force_method == "hybrid":
                intent = QueryIntent.HYBRID
            force_method = ""

        opts = SearchOptions(
            limit=ctx.obj["limit"],
            min_score=ctx.obj["min_score"],
            all=True,
        )

        match intent:
            case QueryIntent.KEYWORD:
                results = store.bm25_search(query, opts)
            case QueryIntent.SEMANTIC:
                print("Vector search requires embedding model setup")
                continue
            case QueryIntent.HYBRID:
                bm25_results = store.bm25_search(query, opts)
                results = store.hybrid_search(bm25_results, [], 60)

        f = Formatter(ctx.obj["format"], ctx.obj["limit"])
        print(f.format_results(results, query))

    print("Goodbye!")

File: qmd-go/pkg/cli/agent.go
package cli

import (
	"bufio"
	"fmt"
	"os"
	"strings"

	"github.com/qmd/qmd-go/pkg/formatter"
	"github.com/qmd/qmd-go/pkg/store"
	"github.com/spf13/cobra"
)

var agentCmd = &cobra.Command{
	Use:   "agent",
	Short: "Run in agent mode (interactive)",
	Run: func(cmd *cobra.Command, args []string) {
		runAgent()
	},
}

// QueryIntent represents the query intent
type QueryIntent string

const (
	IntentKeyword QueryIntent = "keyword"  // BM25
	IntentSemantic QueryIntent = "semantic" // Vector
	IntentHybrid  QueryIntent = "hybrid"    // Hybrid
)

// classifyIntent classifies the query intent
func classifyIntent(query string) QueryIntent {
	// Simple rule-based classification
	lowerQuery := strings.ToLower(query)

	// Check for natural language patterns
	nlPatterns := []string{"explain", "describe", "what is", "how does", "why", "meaning"}
	for _, p := range nlPatterns {
		if strings.Contains(lowerQuery, p) {
			return IntentSemantic
		}
	}

	// Check for technical terms (likely keyword search)
	techPatterns := []string{"error", "exception", "api", "function", "class", "method"}
	for _, p := range techPatterns {
		if strings.Contains(lowerQuery, p) {
			return IntentKeyword
		}
	}

	// Default to hybrid
	return IntentHybrid
}

// runAgent runs the interactive agent
func runAgent() {
	s, err := getStore()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
	defer s.Close()

	fmt.Println("QMD Agent - Interactive Search")
	fmt.Println("Type 'quit' or 'exit' to exit")
	fmt.Println("Type '/bm25', '/vector', or '/hybrid' to force search method")
	fmt.Println()

	reader := bufio.NewReader(os.Stdin)
	forceMethod := ""

	for {
		fmt.Print("> ")
		line, _ := reader.ReadString('\n')
		line = strings.TrimSpace(line)

		if line == "" {
			continue
		}

		if line == "quit" || line == "exit" {
			break
		}

		// Check for forced method
		if strings.HasPrefix(line, "/bm25") {
			forceMethod = "bm25"
			line = strings.TrimPrefix(line, "/bm25 ")
			fmt.Println("Forced BM25 search")
		} else if strings.HasPrefix(line, "/vector") {
			forceMethod = "vector"
			line = strings.TrimPrefix(line, "/vector ")
			fmt.Println("Forced Vector search")
		} else if strings.HasPrefix(line, "/hybrid") {
			forceMethod = "hybrid"
			line = strings.TrimPrefix(line, "/hybrid ")
			fmt.Println("Forced Hybrid search")
		}

		if line == "" {
			continue
		}

		// Classify intent
		intent := classifyIntent(line)
		if forceMethod != "" {
			switch forceMethod {
			case "bm25":
				intent = IntentKeyword
			case "vector":
				intent = IntentSemantic
			case "hybrid":
				intent = IntentHybrid
			}
			forceMethod = ""
		}

		opts := store.SearchOptions{
			Limit:    limit,
			MinScore: minScore,
			All:      true,
		}

		var results []store.SearchResult

		switch intent {
		case IntentKeyword:
			results, _ = s.BM25Search(line, opts)
		case IntentSemantic:
			fmt.Println("Vector search requires embedding model setup")
			continue
		case IntentHybrid:
			results, _ = s.BM25Search(line, opts)
		}

		f := formatter.New(outputFormat, limit)
		fmt.Print(f.FormatSearchResults(results))
	}

	fmt.Println("Goodbye!")
}
